packages/graphql/src/schema-model/relationship/model-adapters/RelationshipAdapter.ts

```diff
@@ -76,7 +82,15 @@ export class RelationshipAdapter {
         if (sourceAdapter) {
             this.source = sourceAdapter;
         } else {
-            this.source = new ConcreteEntityAdapter(source);
+            if (source instanceof ConcreteEntity) {
+                this.source = new ConcreteEntityAdapter(source);
+            } else if (source instanceof InterfaceEntity) {
+                this.source = new InterfaceEntityAdapter(source);
+            } else if (source instanceof UnionEntity) {
+                this.source = new UnionEntityAdapter(source);
+            } else {
+                throw new Error("relationship source must be an Entity");
+            }
```

Comment:
This logic is a bit ugly, I think it may be pointing to a bit of a design issue with how the adapters work tbh :sweat_smile: